### PR TITLE
migrateRealm throws when db doesn't exsit

### LINKED ISSUE #2316: Add a check if database file exists when Realm.migrateRealm() is called.
if `Realm.migrateRealm()` is called when Realm file does not exist, it may create wrong Realm file(no schema contained) or may cause the exception in Migration code.

This issue is reported by @tiro105 on slack chat. Thanks!

We have 2 choices in the case of missing Realm file.

* throw exception if Realm file does not exist
* create Realm file and execute normal initialization of schema instead of executing migration

What do you think? @realm/java

### PR DESCRIPTION
**Pull request by beeender**
Close #2316 .

### CODE REVIEW

File: realm/realm-library/src/main/java/io/realm/Realm.java [hunk 1]
```diff
@@ -1415,8 +1421,9 @@ private void checkMaxDepth(int maxDepth) {
      * version, nothing will happen.
      *
      * @param configuration {@link RealmConfiguration}
+     * @throws FileNotFoundException if the Realm file doesn't exist.
      */
-    public static void migrateRealm(RealmConfiguration configuration) {
+    public static void migrateRealm(RealmConfiguration configuration) throws FileNotFoundException {
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Do we really want to make this a checked exception? I would say it isn't an expected error, but a logical error in the code path?

File: realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java [hunk 2]
```diff
@@ -526,7 +540,7 @@ public void migrate(DynamicRealm realm, long oldVersion, long newVersion) {
     }
 
     @Test
-    public void realmOpenBeforeMigrationThrows() {
+    public void realmOpenBeforeMigrationThrows() throws FileNotFoundException {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java]: See comment below

File: realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java [hunk 3]
```diff
@@ -544,6 +558,29 @@ public void migrate(DynamicRealm realm, long oldVersion, long newVersion) {
...
+        RealmConfiguration config = configFactory.createConfigurationBuilder().migration(new RealmMigration() {
+            @Override
+            public void migrate(DynamicRealm realm, long oldVersion, long newVersion) {
+
+            }
+        }).build();
+        Realm.migrateRealm(config);
+    }
+
+    @Test(expected = FileNotFoundException.class)
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java]: Don't use this annotation if the test case contains multiple lines of code. The wrong code might throw instead: https://github.com/realm/realm-java/blob/master/CONTRIBUTING.md#unit-tests ;)

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 1 above)
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Thinking about this a bit more. FileNotFoundException is a lot better than `IllegalArgument/IllegalState`, and perhaps it is good to force people to consider that the file they try to migrate doesn't exist. I think I could be convinced this is OK :+1:

**Comment by cmelchior** [comment]
Looks good, note that the migration example fails to compile ;)
Apart from small comment about a test case  :+1:

File: realm/realm-library/src/main/java/io/realm/Realm.java (hunk 1 above)
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/Realm.java]: Yeah, I got to the same conclusion after found `FileNotFoundException` is a checked exception. It does make sense to be checked.

**Comment by cmelchior** [comment]
Looks good :+1:

**Comment by stk1m1** [comment]
:+1:

**Comment by zaki50** [comment]
:+1:

File: realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java [hunk 4]
```diff
@@ -544,6 +560,31 @@ public void migrate(DynamicRealm realm, long oldVersion, long newVersion) {
...
+        thrown.expect(FileNotFoundException.class);
+        Realm.migrateRealm(config);
+    }
+
+    @Test
+    public void migrateRealm_configMigration_nonExistingRealmFile() throws FileNotFoundException {
+        RealmMigration migration = new RealmMigration() {
+            @Override
+            public void migrate(DynamicRealm realm, long oldVersion, long newVersion) {
+
```
Comment (studamits) [inline realm/realm-library/src/androidTest/java/io/realm/RealmMigrationTests.java]: zxcv

**Pull request merged**